File: push_sale_mobile-master/lib/models/order.dart
// ignore_for_file: empty_constructor_bodies, non_constant_identifier_names

import 'package:push_sale/models/client.dart';
import 'package:push_sale/models/orderitem.dart';
import 'package:push_sale/models/purchase_order.dart';
import 'package:push_sale/models/tracking_order.dart';

class Order {
  final String id;
  String? actor_id;
  double? total_amount;
  double delivery_amount;
  final String client_id;
  final DateTime order_date;
  String code;
  final DateTime planned_delivery_date;
  DateTime? delivery_date;
  final String state;
  final List<Orderitem> orderitems;
  Client? client;
  List<TrackingOrder>? tracking;
  List<PurchaseOrder>? purchase_orders;
  Order(
      {required this.id,
      this.actor_id,
      this.code = "",
      this.total_amount,
      this.delivery_amount = 0,
      required this.client_id,
      required this.order_date,
      required this.planned_delivery_date,
      required this.delivery_date,
      required this.state,
      required this.orderitems,
      this.client,
      this.tracking,
      this.purchase_orders});

  Map<String, dynamic> toMap() {
    Map<String, dynamic> ret = {};
    ret["id"] = id;
    ret["client_id"] = client_id;
    ret["order_date"] = order_date.toString();
    ret["planned_delivery_date"] = planned_delivery_date.toString();
    ret["state"] = state;
    ret["orderitems"] = orderitems.map((e) => e.toMap()).toList();
    return ret;
  }

  static Order fromMap(Map<String, dynamic> value) {
    return Order(
      id: value["id"],
      client_id: value["client_id"],
      total_amount: double.parse(value["total_amount"].toString()),
      delivery_amount: value["purchase_orders_sum_total_amount"] != null
          ? double.parse(value["purchase_orders_sum_total_amount"].toString()) -
              double.parse(value["purchase_orders_sum_residual"].toString())
          : 0,
      code: value["code"],
      order_date: DateTime.parse(value["order_date"]),
      planned_delivery_date: DateTime.parse(value["planned_delivery_date"]),
      delivery_date: value["delivery_date"] != null
          ? DateTime.parse(value["delivery_date"])
          : null,
      state: value["state"],
      orderitems: Orderitem.fromListMapToList(value["orderitem"] ?? []),
      client: value["client"] != null ? Client.fromMap(value["client"]) : null,
      tracking: value["tracking"] != null
          ? TrackingOrder.fromListMapToList(value["tracking"])
          : null,
    );
  }

  static Order fromMapStatus(Map<String, dynamic> value) {
    return Order(
        id: value["id"],
        client_id: value["client_id"],
        total_amount: double.parse(value["total_amount"].toString()),
        delivery_amount: value["purchase_orders_sum_total_amount"] != null
            ? double.parse(
                    value["purchase_orders_sum_total_amount"].toString()) -
                double.parse(value["purchase_orders_sum_residual"].toString())
            : 0,
        code: value["code"],
        order_date: DateTime.parse(value["order_date"]),
        planned_delivery_date: DateTime.parse(value["planned_delivery_date"]),
        delivery_date: value["delivery_date"] != null
            ? DateTime.parse(value["delivery_date"])
            : null,
        state: value["state"],
        orderitems: Orderitem.fromListMapToList(value["orderitem"] ?? []),
        client:
            value["client"] != null ? Client.fromMap(value["client"]) : null,
        tracking: value["tracking"] != null
            ? TrackingOrder.fromListMapToList(value["tracking"])
            : null,
        purchase_orders:
            PurchaseOrder.fromListMapToList(value["purchase_orders"]));
  }
}

File: push_sale-master/push_sale-master/app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Laravel\Passport\HasApiTokens;
use Exception;

class Order extends Model
{
    use HasApiTokens, HasFactory;
    protected $table = "order";

    public $incrementing = false;
    protected $keyType = 'string';

    protected $fillable = ["id", "code", "actor_id", "client_id", "total_amount", "residual", "order_date", "planned_delivery_date", "state"];

    public function orderitem()
    {
        return $this->hasMany(OrderItem::class, "order_id", "id");
    }

    public function client()
    {
        return $this->hasOne(Client::class, "id", "client_id");
    }

    public function tracking()
    {
        return $this->hasMany(TrackingOrders::class, "order_id", "id");
    }

    public function PurchaseOrders(){
        return $this->hasMany(PurchaseOrder::class, "order_id");
    }

    public function toPurchaseOrder($track_id)
    {
        try {
            $purchaseOrders = [];
            $total = [];
            $i = 0;
            $j = 0;
            foreach ($this->orderitem as $item) {
                if (!isset($purchaseOrders[$item->warehouse_id])) {
                    $purchaseOrders[$item->warehouse_id] = new PurchaseOrder();
                    $purchaseOrders[$item->warehouse_id]->id = $this->id . "-" . $i;
                    $purchaseOrders[$item->warehouse_id]->code = $this->code;
                    $purchaseOrders[$item->warehouse_id]->order_id = $this->id;
                    $purchaseOrders[$item->warehouse_id]->actor_id = null;
                    $purchaseOrders[$item->warehouse_id]->client_id = $this->client_id;
                    $purchaseOrders[$item->warehouse_id]->type = "invoice_out";
                    $purchaseOrders[$item->warehouse_id]->warehouse_id = $item->warehouse_id;
                    $purchaseOrders[$item->warehouse_id]->total_amount = 0;
                    $purchaseOrders[$item->warehouse_id]->purchase_date = $this->order_date;
                    $purchaseOrders[$item->warehouse_id]->planned_delivery_date = $this->planned_delivery_date;
                    $purchaseOrders[$item->warehouse_id]->delivery_date = null;
                    $purchaseOrders[$item->warehouse_id]->state = "new";

                    $total[$purchaseOrders[$item->warehouse_id]->id] = 0;
                    $i++;
                }
                $purchaseItem = new PurchaseOrderItem();
                $purchaseItem->id = $item->id . "-" . $j;
                $purchaseItem->purchaseorder_id = $purchaseOrders[$item->warehouse_id]->id;
                $purchaseItem->image = $item->image;
                $purchaseItem->product_name = $item->product_name;
                $purchaseItem->variant_name_1 = $item->variant_name_1;
                $purchaseItem->variant_name_2 = $item->variant_name_2;
                $purchaseItem->promotion_id = $item->promotion_id;
                $purchaseItem->promotionitem_id = $item->promotionitem_id;
                $purchaseItem->coupon_id = $item->coupon_id;
                $purchaseItem->unite = $item->unite;
                $purchaseItem->discount = $item->discount;
                $purchaseItem->variant_id = $item->variant_id;
                $purchaseItem->sku = $item->sku;
                $purchaseItem->quantity = $item->quantity;
                $purchaseItem->package = $item->package;
                $purchaseItem->price = $item->price;

                $total[$purchaseOrders[$item->warehouse_id]->id] += $item->price * $item->quantity;

                $purchaseItem->save();

                // $purchaseOrders[$item->warehouse_id]->orderitem()->add($purchaseItem);
                $j++;
            }
            $po = 0;
            foreach ($purchaseOrders as $purchaseOrder) {
                $purchaseOrder->total_amount = $total[$purchaseOrder->id];
                $purchaseOrder->residual = $total[$purchaseOrder->id];
                $track = TrackingOrders::where("id", $track_id . "-" . $po)->first();
                if (!$track) {
                    TrackingOrders::create([
                        "id" => $track_id . "-" . $po,
                        "order_id" => $this->id,
                        "purchaseorder_id" => $purchaseOrder->id,
                        "state" => "new",
                        "amount" => $purchaseOrder->total_amount,
                        "actor_id" => $this->actor_id,
                        "is_last" => false,
                    ]);
                    $purchaseOrder->save();
                }
            }
            return "done";
        } catch (Exception $e) {
            return $e->getMessage();
        }
    }
}
